Add default HKDF expander implementation

Ryanfsdf opened the pull request:
Added an implementation of the HKDF interface. I don't like the idea of calling it `default` but I couldn't think of anything else to call it. Any thoughts?

Also, I couldn't decide if I should put the implementation in the same file as the interface or create a separate file for it (`defaulthkdf.go`) or keep it in the `halfconn` file as I have it right now. I'll be adding the Half Connection code to the `halfconn` file in a following PR. Any thoughts on this as well?

matthewstevenson88 commented on security/s2a/internal/crypter/halfconn.go:
```diff
@@ -0,0 +1,28 @@
+	"hash"
+)
+
+// defaultHKDFExpander is the default HKDF expander which uses Go's crypto/hkdf
```
s/expander/expansion function,

Ryanfsdf replied:
I think keeping expander here might be better than swapping it with expansion function. The class itself is the expander class and its main purpose is to provide an expansion function method. I think if we swap expander for expansion function, we would be implying that the class itself is the expansion function, which might be a bit confusing. WDYT?

matthewstevenson88 replied:
Sounds reasonable to me. Let's leave it. :)

matthewstevenson88 commented:
Thanks for the PR Ryan! It looks good overall. Two broad comments:

- I think it makes more sense to have this code in `expander.go`, and then everything else can go in `halfconn.go`. (If you look in ALTS's [`record.go`](https://github.com/grpc/grpc-go/blob/4346c5926defb126b3fadee6c46308fc746be3f2/credentials/alts/internal/conn/record.go), they have an interface and its implementation in the same file, so doing so isn't against Go style.)
-Can you add a small test that uses the test vectors from [RFC 5869](https://tools.ietf.org/html/rfc5869#appendix-A)? Only use the 3 test vectors that use SHA-256, since we don't want to use SHA-1. (Note that you will only need 4 of the entries from the test vector, namely the PRK, info, length, and OKM.)

(Also, have you run `gofmt`? I feel like the declaration of `expand` looks strangely formatted, but it might just be Github or my browser.)

Ryanfsdf commented:
Thanks for the review! I added the test cases and moved the implementation over to `expander.go`. PTAL!

Also, yes I ran `gofmt`, I have my IDE configured to run it every time I save a file. I think it looks properly formatted, I think it might be how github shows the code.

matthewstevenson88 commented on security/s2a/internal/crypter/expander.go:
```diff
@@ -12,3 +14,24 @@ type hkdfExpander interface {
+	outBuf := make([]byte, length)
+	n, err := hkdf.Expand(hash, secret, label).Read(outBuf)
+	if err != nil {
+		return nil, fmt.Errorf("hkdf.expand.Read failed with error: %v", err)
```
Sorry I forgot to mention this in my last round of comments. Could we also add a test case to hit this return statement?

Ryanfsdf replied:
Done

matthewstevenson88 commented:
Thanks Ryan! Looks good to me, aside from on test case I forgot to mention before. When that's added, let me know and I'll approve.

matthewstevenson88 commented:
Thanks for the PR Ryan! LGTM. I'm adding @cesarghali for Go readability.

cesarghali commented on security/s2a/internal/crypter/expander.go:
```diff
@@ -12,3 +14,24 @@ type hkdfExpander interface {
 	expand(hash func() hash.Hash, secret, label []byte, length int) ([]byte, error)
 }
+
+// defaultHKDFExpander is the default HKDF expander which uses Go's crypto/hkdf
```
The use of the term "default" implies that there might be a non-default HKDF expander, but I don't think that's the case. Maybe call it expander?

Ryanfsdf replied:
If we ever use a separate crypto library for HKDF expansion like @matthewstevenson88 mentioned, I think it'll probably be called `<library_name>HKDFExpander`. Keeping this name as `expander` might be confusing if we add these separate crypto libraries later. I called it default because it uses Go's own crypto library for HKDF expansion.  WDYT?

cesarghali replied:
SG, thanks for the explanation.

cesarghali commented on security/s2a/internal/crypter/expander.go:
```diff
@@ -12,3 +14,24 @@ type hkdfExpander interface {
+	return &defaultHKDFExpander{}
+}
+
+func (*defaultHKDFExpander) expand(hash func() hash.Hash, secret, label []byte, length int) ([]byte, error) {
```
Thinking about this API again. Is the output length related to the returned hash function resulting from calling `hash`? If that's the case, can we get the `length` from that?

Ryanfsdf replied:
I believe we need to pass in `length` for efficiency sake. `hkdf.Expand` returns an `io.Reader` interface which exposes only a `Read(p []byte) (n int, err error)` method. We need to pass in a correctly allocated buffer here so that we can do 1 buffer allocation and 1 read rather than numerous small reads. I don't believe it's possible to get the `length` directly from the resulting call of `hkdf.Expand`. If we want to get the size directly from the hash function, we need to do `expander := hmac.New(hash, secret).Size()` before we allocate the buffer and call `hkdf.Expand`, which seems inefficient. I think it's best to leave the `length` input as it is. Please correct if I'm wrong!

cesarghali replied:
Calling `hash` returns a `hash.Hash` object [1] which has a `Size` API. There's no need to create an HMAC object to get the size of the underlying hash function. You can simply do:

```
length := hash().Size()
```

[1] https://golang.org/pkg/hash/#Hash

cesarghali replied:
Also, maybe we should rename the `hash` parameter to `h` to differentiate that from the imported `hash` package. We should do the same in the interface as well. Feel free to do that in this PR.

Ryanfsdf replied:
Done.

cesarghali commented on security/s2a/internal/crypter/expander.go:
```diff
@@ -12,3 +14,24 @@ type hkdfExpander interface {
+		return nil, fmt.Errorf("hkdf.expand.Read failed with error: %v", err)
+	}
+	if n != length {
+		return nil, fmt.Errorf("hkdf.expand.Read returned unexpected length, got: %d, want: %d", n, length)
```
To me it seems that this must not happen. If the caller provided the right `hash` with the right `length`, then `n` must be equal to `length`. If that's the case, then maybe we should `panic` here. However, we might not need this check at all. See my comment above.

Ryanfsdf replied:
Looking through the implementation of `Read` for an `hdkf` object, I don't think it's possible to ever return an invalid length while not returning an error. Since we check for an error in the line above, I don't think its possible to hit this condition.

However, I think the documentation of the `Read` interface suggests that other implementations may return a nil error while still returning a number that is less than the number that was requested for. As you mentioned though, if we provide the right `hash` with the right `length`, I think hitting this condition may signal a programming error and so `panic` may be the correct here.

Ryanfsdf replied:
Since `length` is no longer a parameter, I don't think we should `panic` here. I left it as an error for now, I don't know if it's worth doing the check at all however.

cesarghali commented on security/s2a/internal/crypter/expander_test.go:
```diff
@@ -0,0 +1,66 @@
+	}
+}
+
+func TestExpanderExpansion(t *testing.T) {
```
If a test is testing a specific function, it's a good to call the test `TestFuncName`. In this case, this can be called `TestExpand`.

Ryanfsdf replied:
Done.

cesarghali commented on security/s2a/internal/crypter/expander_test.go:
```diff
@@ -0,0 +1,66 @@
+	length         int
+}
+
+func TestExpanderExpansionInvalidLength(t *testing.T) {
```
This test is not needed to be separate. Its test case can be merged with `TestExpand`. You might want to add a boolean `success` to the test vector to know if the test case is support to fail or succeed.

Ryanfsdf replied:
This test is no longer valid since `length` is no longer a parameter so I removed it altogether.

cesarghali commented on security/s2a/internal/crypter/expander_test.go:
```diff
@@ -0,0 +1,66 @@
+
+func TestExpanderExpansion(t *testing.T) {
+	// Test vectors were taken from https://tools.ietf.org/html/rfc5869.
+	for _, tc := range []hkdfTestVector{
```
Since the test vector struct is only used here, there's no need to define it separately. Instead you can do:

```
for _, tc := range []struct{
	comment        string
	prk, info, okm []byte
	length         int
}{
	{
		comment: ...,
		prk: ...,
	},
	...
}{
	// the test code.
}
```

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/expander_test.go:
```diff
@@ -0,0 +1,66 @@
+)
+
+type hkdfTestVector struct {
+	comment        string
```
Let's call this `desc`.

Also, sorry for missing that in the PR #8. Can you please send a separate PR to change `comment` to `desc`?

Ryanfsdf replied:
Done. Will change the rest in a separate PR.

cesarghali commented on security/s2a/internal/crypter/expander_test.go:
```diff
@@ -0,0 +1,66 @@
+				t.Errorf("expand failed with error: %v", err)
+			}
+			if !bytes.Equal(got, tc.okm) {
+				t.Errorf("expand(sha256.New, %v, %v, %v) = %v, want: %v.", tc.prk, tc.info, tc.length, got, tc.okm)
```
No need for `:` after `want`. Please make sure this is the case everywhere. If other tests need to be adjusted, let's do that in a separate PR.

Ryanfsdf replied:
Done. Will change the rest in a separate PR.

cesarghali commented on security/s2a/internal/crypter/expander.go:
```diff
@@ -24,14 +24,15 @@ func newDefaultHKDFExpander() hkdfExpander {
 	}
 	if n != length {
-		return nil, fmt.Errorf("hkdf.expand.Read returned unexpected length, got: %d, want: %d", n, length)
+		return nil, fmt.Errorf("hkdf.expand.Read returned unexpected length, got %d, want %d", n, length)
```
s/expand/Expand

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/expander.go:
```diff
@@ -24,14 +24,15 @@ func newDefaultHKDFExpander() hkdfExpander {
 	if err != nil {
 		return nil, fmt.Errorf("hkdf.expand.Read failed with error: %v", err)
 	}
 	if n != length {
```
`Read` is guaranteed not to return `n > length`. Should we change this check to `n < length` explicitly?

cesarghali replied:
It's fine if you want to keep it as is, but please add a comment explaining my comment here, if it is correct.

Ryanfsdf replied:
You're correct. I think it's better to check `n < length` explicitly. Done.

cesarghali commented on security/s2a/internal/crypter/expander_test.go:
```diff
@@ -7,59 +7,41 @@ import (
+func TestExpand(t *testing.T) {
+	// The following test vectors were taken from
+	// https://tools.ietf.org/html/rfc5869. Note that they have been
+	// slightly modified to fit our interface and implementation.
```
"slightly modified" is a big vague and for test vectors taken from the RFC, it's a bit scary :)

I suggest mention in the comment how the test vectors are exactly modified.

Ryanfsdf replied:
Done.

cesarghali commented on security/s2a/internal/crypter/expander_test.go:
```diff
@@ -9,10 +9,12 @@ import (
-	// and `out`.
+	// https://tools.ietf.org/html/rfc5869. Note have vectors have
+	// been slightly modified to test our specific implementation. In
+	// particular, the output has been shortened since we our implementation
```
s/shortened/truncated to a specific length

Ryanfsdf replied:
Done

Ryanfsdf commented:
Thanks for all the reviews!
